src/metabase/driver/sql_jdbc/execute.clj

```diff
@@ -329,29 +330,27 @@
         (.close stmt)
         (throw e)))))
 
-(defn- prepared-statement*
-  ^PreparedStatement [driver conn sql params canceled-chan]
-  ;; if canceled-chan gets a message, cancel the PreparedStatement
-  (let [^PreparedStatement stmt (prepared-statement driver conn sql params)]
+(defn- wire-up-canceled-chan-to-cancel-Statement!
+  "If `canceled-chan` gets a message, cancel the Statement `stmt`."
+  [^Statement stmt canceled-chan]
+  (when canceled-chan
     (a/go
       (when (a/<! canceled-chan)
-        (log/debug (trs "Query canceled, calling PreparedStatement.cancel()"))
+        (log/debug (trs "Query canceled, calling Statement.cancel()"))
         (u/ignore-exceptions
-          (.cancel stmt))))
-    stmt))
+          (.cancel stmt))))))
+
+(defn- prepared-statement*
+  ^PreparedStatement [driver conn sql params canceled-chan]
+  (doto (prepared-statement driver conn sql params)
+    (wire-up-canceled-chan-to-cancel-Statement! canceled-chan)))
 
 (defn- use-statement? [driver params]
   (and (statement-supported? driver) (empty? params)))
 
 (defn- statement* ^Statement [driver conn canceled-chan]
-  ;; if canceled-chan gets a message, cancel the Statement
-  (let [^Statement stmt (statement driver conn)]
-    (a/go
-      (when (a/<! canceled-chan)
-        (log/debug (trs "Query canceled, calling Statement.cancel()"))
-        (u/ignore-exceptions
-         (.cancel stmt))))
-    stmt))
+  (doto (statement driver conn)
+    (wire-up-canceled-chan-to-cancel-Statement! canceled-chan)))
```

Comment:
Deduplicated this code a bit and made it work if `canceled-chan` is `nil` as it is with the new writeback code path